src/azure-cli/azure/cli/command_modules/vm/custom.py

```diff
@@ -348,18 +348,73 @@ def create_managed_disk(cmd, resource_group_name, disk_name, location=None,  # p
                 if disk_version.lower() == 'latest':
                     disk_version = _get_latest_image_version(cmd.cli_ctx, location, disk_publisher, disk_offer,
                                                              disk_sku)
-                client = _compute_client_factory(cmd.cli_ctx)
-                response = client.virtual_machine_images.get(location, disk_publisher, disk_offer, disk_sku,
-                                                             disk_version)
-                image_reference = response.id
             else:  # error
                 raise CLIError('usage error: --image-reference should be ID or URN (publisher:offer:sku:version).')
+        else:
+            from msrestazure.tools import parse_resource_id
+            terms = parse_resource_id(image_reference)
+            disk_publisher, disk_offer, disk_sku, disk_version = \
+                terms['child_name_1'], terms['child_name_3'], terms['child_name_4'], terms['child_name_5']
+
+        client = _compute_client_factory(cmd.cli_ctx)
+        response = client.virtual_machine_images.get(location, disk_publisher, disk_offer, disk_sku,
+                                                     disk_version)
+
+        if hasattr(response, 'hyper_v_generation') and response.hyper_v_generation == 'V2' and not image_reference_lun:
+            # will set default value of hyper_v_generation
+            logger.warning('Starting Build 2023 event, az disk create command '
+                           'will deploy Trusted Launch VM by default. To know more about Trusted Launch, '
+                           'please visit https://docs.microsoft.com/en-us/azure/virtual-machines/trusted-launch')
```

Comment:
I guess this should be a warning log printed for Gen 1 disk
```suggestion
        if hasattr(response, 'hyper_v_generation') and response.hyper_v_generation == 'V1' and not image_reference_lun:
            # will set default value of hyper_v_generation
            logger.warning('Starting Build 2023 event, az disk create command '
                           'will deploy Trusted Launch VM by default. To know more about Trusted Launch, '
                           'please visit https://docs.microsoft.com/en-us/azure/virtual-machines/trusted-launch')
```

Reply:
@AjKundnani May I ask in the warning log, do we need to consider recommending users to use the V2 hypervisor generation?

Reply:
@AjKundnani In addition, we may need to consider recommending users to use specific parameters `--hyper-v-generation V2` and `--security-type TrustedLaunch` in the log to change their existing behavior, otherwise users are not sure how to handle this warning. What's your opinion?

Reply:
> @AjKundnani May I ask in the warning log, do we need to consider recommending users to use the V2 hypervisor generation?

@zhoxing-ms - No, Trusted Launch VMs require V2 hypervisor generation, it will not work with V1 due to UEFI requirements.

Reply:
> No, Trusted Launch VMs require V2 hypervisor generation, it will not work with V1 due to UEFI requirements.

Yes, I know that. I mean, when users use the Gen 1 image, do we need a warning log to recommend them to use Gen 2 image?

Reply:
> > No, Trusted Launch VMs require V2 hypervisor generation, it will not work with V1 due to UEFI requirements.
> 
> Yes, I know that. I mean, when users use the Gen 1 image, do we need a warning log to recommend them to use Gen 2 image?

- Is it feasible to query Gen2 equivalent image for user provided Gen1 input? For example, if user provides `2022-Datacenter` then CLI is able to query and recommend `2022-datacenter-azure-edition-core`?
- Alternatively, if we can point users to Trusted Launch documentation using warning message discussed above - `Consider upgrading security for your VM resources using Azure Trusted Launch VMs. To know more about Trusted Launch, please visit https://docs.microsoft.com/en-us/azure/virtual-machines/trusted-launch`

Reply:
@AjKundnani We found that when we create disk using Gen2 image and --hyper-v-generation set to V1, no error is reported. Thus, we wonder if we need a warning log to users, which is mentioned above by @zhoxing-ms. Attached below are Gen2 image and the output of creating disk.
![8a4008b72d3ac1b722043c5589d707a](https://user-images.githubusercontent.com/105691024/191476775-df085683-a442-4d58-99f8-f6f1692d6352.png)
![db186a649c0f5590438464f8eecb6df](https://user-images.githubusercontent.com/105691024/191476821-a0f7ff61-b6b1-4a40-9cee-4197e1460c3a.png)

Reply:
> @AjKundnani We found that when we create disk using Gen2 image (**--image-reference is set to MicrosoftWindowsServer:WindowsServer:2022-datacenter-smalldisk-g2:latest**) and **--hyper-v-generation set to V1**, no error is reported. Thus, we wonder if we need a warning log to prompt users to use `--hyper-v-generation V2` for Gen 2 image, which is mentioned above by @zhoxing-ms. Attached below are Gen2 image and the output of creating disk. ![8a4008b72d3ac1b722043c5589d707a](https://user-images.githubusercontent.com/105691024/191476775-df085683-a442-4d58-99f8-f6f1692d6352.png) ![db186a649c0f5590438464f8eecb6df](https://user-images.githubusercontent.com/105691024/191476821-a0f7ff61-b6b1-4a40-9cee-4197e1460c3a.png)

@yanzhudd - Ideally this should be reported as bug with API team, I will try this scenario and file the bug. My thought is CLI should pass user inputs as-is to API. In this scenario, if user has passed image but did not use parameter `--hyper-v-generation`, the value should default to **V2** as per request.

Reply:
> MicrosoftWindowsServer:WindowsServer:2022-datacenter-smalldisk-g2:lates

@yanzhudd - I am unable to recreate this scenario, I can share debug log over teams if needed.

Command Used `az disk create -g cli-disk-rg -n cliosdisk --image-reference MicrosoftWindowsServer:WindowsServer:2022-datacenter-smalldisk-g2:latest --hyper-v-generation v1`

Output

```
{
  "burstingEnabled": null,
  "completionPercent": null,
  "creationData": {
    "createOption": "FromImage",
    "galleryImageReference": null,
    "imageReference": {
      "communityGalleryImageId": null,
      "id": "/Subscriptions/821664e1-b43e-4312-a018-4d2d284c2b9c/Providers/Microsoft.Compute/Locations/WestUS3/Publishers/MicrosoftWindowsServer/ArtifactTypes/VMImage/Offers/WindowsServer/Skus/2022-datacenter-smalldisk-g2/Versions/20348.1006.220908",
      "lun": null,
      "sharedGalleryImageId": null
    },
    "logicalSectorSize": null,
    "securityDataUri": null,
    "sourceResourceId": null,
    "sourceUniqueId": null,
    "sourceUri": null,
    "storageAccountId": null,
    "uploadSizeBytes": null
  },
  "dataAccessAuthMode": null,
  "diskAccessId": null,
  "diskIopsReadOnly": null,
  "diskIopsReadWrite": 120,
  "diskMBpsReadOnly": null,
  "diskMBpsReadWrite": 25,
  "diskSizeBytes": 32214351872,
  "diskSizeGb": 30,
  "diskState": "Unattached",
  "encryption": {
    "diskEncryptionSetId": null,
    "type": "EncryptionAtRestWithPlatformKey"
  },
  "encryptionSettingsCollection": null,
  "extendedLocation": null,
  "hyperVGeneration": "V2",
  "id": "/subscriptions/821664e1-b43e-4312-a018-4d2d284c2b9c/resourceGroups/cli-disk-rg/providers/Microsoft.Compute/disks/cliosdisk",
  "location": "westus3",
  "managedBy": null,
  "managedByExtended": null,
  "maxShares": null,
  "name": "cliosdisk",
  "networkAccessPolicy": "AllowAll",
  "osType": "Windows",
  "propertyUpdatesInProgress": null,
  "provisioningState": "Succeeded",
  "publicNetworkAccess": "Enabled",
  "purchasePlan": null,
  "resourceGroup": "cli-disk-rg",
  "securityProfile": null,
  "shareInfo": null,
  "sku": {
    "name": "Premium_LRS",
    "tier": "Premium"
  },
  "supportedCapabilities": {
    "acceleratedNetwork": true,
    "architecture": "x64"
  },
  "supportsHibernation": true,
  "tags": {},
  "tier": "P4",
  "timeCreated": "2022-09-22T11:12:30.868005+00:00",
  "type": "Microsoft.Compute/disks",
  "uniqueId": "770f6242-a579-4369-82c2-1153cde23ec6",
  "zones": null
}
```

Reply:
> 

Hi @AjKundnani Maybe this question has been fixed. I now run out is also 'V2'.

Reply:
> The OS image will support only one of Gen i.e. it'll be either Gen1 (v1) or Gen 2 (v2) but not both. (Unlike VM Size which can support both v1 and v2)

@AjKundnani OK, you were right 👍

> Alternatively, if we can point users to Trusted Launch documentation using warning message discussed above - Consider upgrading security for your VM resources using Azure Trusted Launch VMs. To know more about Trusted Launch, please visit https://docs.microsoft.com/en-us/azure/virtual-machines/trusted-launch

@AjKundnani Do you think we need to recommend the Gen 2 image to users in the warning log when they are using the Gen 1 image? such as: Please consider upgrading security for your VM resources by using **Gen 2 image** and Trusted Launch VMs. To know more about Trusted Launch, please visit https://docs.microsoft.com/en-us/azure/virtual-machines/trusted-launch

Reply:
> Please consider upgrading security for your VM resources by using **Gen 2 image** and Trusted Launch VMs. To know more about Trusted Launch, please visit https://docs.microsoft.com/en-us/azure/virtual-machines/trusted-launch

`Please consider upgrading security for your VM resources by using Gen 2 OS image and Trusted Launch security type. To know more about Trusted Launch, please visit https://docs.microsoft.com/en-us/azure/virtual-machines/trusted-launch`